## **Introduction**
Hey! Welcome to the analysis of the Shipping Problem!
My name is

 $\text{Kartik Srinivas}$


$\text{Roll  Number :ES20BTECH11015}$

The problem has been divided into 3 parts and solved, all code blocks are executed and the necessary equations have been written in $\text{LaTeX}$



## **The Data**
The data of the problem is very straightforward, the notation used for matrices are as follows:

1.   $Demand = D \text{   } \epsilon \text{ }\mathbb{R}^{{5}\times{1}}$
2.   $Cost = C \text{   } \epsilon \text{ }\mathbb{R}^{{3}\times{5}}$
1.   $Supply = S \text{   } \epsilon \text{ }\mathbb{R}^{{3}\times{5}}$





In [1]:
import cvxpy as cp
import numpy as np

d = np.matrix([45,20,30,30,10]).T
c = np.matrix([[8,6,10,9,8],[9,12,13,7,5],[14,9,16,5,2]])
s = np.matrix([40,50,45]).T
n = 3;
m = 5;

## **The Variables and Constraints**
The Variable $V$ has been defined as an $\mathbb{R}^{{3}\times{5}}$ matrix. We denote it as follows

$V _{ij}$ = $\text{Amount of goods shipped from location } i \text{ to factory } j  $

The constraints are as follows:


*   Only a positive Amount can be shifted, so we have $V_{ij}>=0$
*   Sum of total Shipment from a factory cannot exceed supply so we get
$∑_{j = 1}^{j = 5}V_{ij}  <= S_{i}$

*   The demand Must be satisfied $∑_{i = 1}^{i = 3}V_{ij} > = D_i$




In [2]:
v = cp.Variable((n,m));
print("shapes =",np.shape(d), np.shape(c), np.shape(s))
constraints = [v>=0, v.T@np.ones((n,1)) >= d, v@np.ones((m,1))  ==s];

shapes = (5, 1) (3, 5) (3, 1)


## **Optimization**
The Optimization is to be done on the total cost.


The total cost can be seen as the element wise product of the cost matrix$(C)$ and the amount matrix $(V)$

$Cost = \displaystyle ∑_{i}∑_{j} V_{ij} \times C_{ij}$

In [4]:
objective = cp.Minimize( cp.sum((cp.multiply(c,v)))) #this is element wise multiplication
prob = cp.Problem(objective,constraints);
result = prob.solve();
print("Optimal Cost result = \n",result)
#for line in v.value:
    #print('  '.join(map(str, line)))
print("Optimal Giveaway matrix= \n")
for i in range(3):
    for j in range(5):
        print(v.value[i][j],end = "|")
    print("\n")


Optimal Cost result = 
 1025.0000000255045
Optimal Giveaway matrix= 

6.151022161968637e-10|14.999999999862935|24.999999998993612|4.749702763345747e-10|5.637317186374904e-11|

44.99999999837126|1.999878668806543e-09|4.999999998236501|7.523169778691479e-10|6.431943139683385e-10|

9.419988439622996e-10|4.99999999787936|2.805678159289452e-09|29.99999999870076|9.999999999675351|



## **Result**
The final result yields an optimal cost of $1025 {$} $ and the Entries are also positive. This completes the problem